lib/view_component/compiler.rb

```diff
@@ -103,266 +50,273 @@ def renders_template_for?(variant, format)
-            .select { |_, count| count > 1 }
-            .each do |tally|
-            variant, this_format = tally[0]
+        variant_string = " for variant `#{variant}`" if variant.present?
 
-            variant_string = " for variant `#{variant}`" if variant.present?
+        errors << "More than one #{this_format.upcase} template found#{variant_string} for #{@component}. "
+      end
 
-            errors << "More than one #{this_format.upcase} template found#{variant_string} for #{component_class}. "
-          end
+      if @templates.any? { _1.variant.nil? && !_1.inline_call? } &&
+          @templates.any? { _1.variant.nil? && _1.inline_call? && _1.defined_on_self? }
 
-          if templates.find { |template| template[:variant].nil? } && inline_calls_defined_on_self.include?(:call)
-            errors <<
-              "Template file and inline render method found for #{component_class}. " \
-              "There can only be a template file or inline render method per component."
-          end
+        errors <<
+          "Template file and inline render method found for #{@component}. " \
+          "There can only be a template file or inline render method per component."
+      end
 
-          duplicate_template_file_and_inline_variant_calls =
-            templates.pluck(:variant) & variants_from_inline_calls(inline_calls_defined_on_self)
+      duplicate_template_file_and_inline_call_variants =
+        @templates.select { !_1.inline_call? }.map(&:variant) &
+        @templates.select { _1.inline_call? && _1.defined_on_self? }.map(&:variant)
 
-          unless duplicate_template_file_and_inline_variant_calls.empty?
-            count = duplicate_template_file_and_inline_variant_calls.count
+      unless duplicate_template_file_and_inline_call_variants.empty?
+        count = duplicate_template_file_and_inline_call_variants.count
 
-            errors <<
-              "Template #{"file".pluralize(count)} and inline render #{"method".pluralize(count)} " \
-              "found for #{"variant".pluralize(count)} " \
-              "#{duplicate_template_file_and_inline_variant_calls.map { |v| "'#{v}'" }.to_sentence} " \
-              "in #{component_class}. " \
-              "There can only be a template file or inline render method per variant."
-          end
+        errors <<
+          "Template #{"file".pluralize(count)} and inline render #{"method".pluralize(count)} " \
+          "found for #{"variant".pluralize(count)} " \
+          "#{duplicate_template_file_and_inline_call_variants.map { |v| "'#{v}'" }.to_sentence} " \
+          "in #{@component}. There can only be a template file or inline render method per variant."
+      end
 
-          uniq_variants = variants.compact.uniq
-          normalized_variants = uniq_variants.map { |variant| normalized_variant_name(variant) }
+      variant_pairs =
+        @templates.select { _1.variant.present? }.map { [_1.variant, _1.normalized_variant_name] }.uniq { _1.first }
```

Comment:
```suggestion
        @templates.each_with_object({}) do |template, pairs|
          next unless template.variant
          next if pairs[template.variant]
          
          pairs[template.variant] = template.normalized_variant_name
        end
```

I think this is equivalent, but avoids a few extra loops (I also omitted `.present?` too since I often consider a lot of this code "performance" Ruby)